tools/x-repo-types/runner/typeAnalyzer.go

```diff
@@ -0,0 +1,459 @@
+	Type     reflect.Type
+	Kind     reflect.Kind
+	Edges    []Edge
+	children *Children
+}
+
+// String returns fully qualified information about the Type.
+func (t *Type) String() string {
+	return fmt.Sprintf("%s :: %q (%s)", t.Type.PkgPath(), t.Type, t.Kind)
+}
+
+// Children represent a Type's child Types.
+type Children map[string]Type
+
+// Edge represents the name a type is referred together with
+// the Tag metadata (for the case of a struct field)
+type Edge struct {
+	Name, Tag string
+}
+
+// String represents the Edge as in an invertible manner.
+func (e *Edge) String() string {
+	return fmt.Sprintf("[%s](%s)", e.Name, e.Tag)
+}
+
+// EdgeFromLabel de-"serialize" and Edge that has be rendered via Edge.String()
+func EdgeFromLabel(s string) *Edge {
+	re := regexp.MustCompile(`^\[(.+)\]\((.+)\)$`)
+	matches := re.FindStringSubmatch(s)
+	if len(matches) == 3 {
+		return &Edge{Name: matches[1], Tag: matches[2]}
+	}
+	return nil
+}
+
+// Target represents a Type that is a child of another Type by
+// providing the Edge that "points" to it.
+// In the case of a "root" Type, it's up to the caller to provide
+// the Edge.
+type Target struct {
+	Edge
+	Type Type
+}
+
+// String is a convenience method for printing a Target.
+func (tgt *Target) String() string {
+	return fmt.Sprintf("%s|-->%s", &tgt.Edge, &tgt.Type)
+}
+
+// SerializationInfo provides the essential data
+// for go to serialize the field or other type.
+func (e *Edge) SerializationInfo() string {
+	// Probably more subtlety is called for.
+	re := regexp.MustCompile(`^codec:"([^,"]+)`)
+
+	if e.Tag == "" {
+		return e.Name
+	}
+
+	matches := re.FindStringSubmatch(e.Tag)
+	if len(matches) > 1 {
+		return matches[1]
+	}
+	return e.Tag
+}
+
+// TargetPair represents a pair of Targets.
+type TargetPair struct {
+	X, Y Target
+}
+
+// Diff represents a difference (if any) between two Types.
+// CommonPath is the path in the trees of the types from their roots
+// that arrives at the difference.
+// Xdiff and Ydiff are the differences at the end of the common path
+// between children. Set theoretically:
+// * Xdiff = X - Y
+// * Ydiff = Y - X.
+type Diff struct {
+	CommonPath   []TargetPair
+	Xdiff, Ydiff []Target
+}
+
+// Empty reports whethere there was any difference at all.
+func (d *Diff) Empty() bool {
+	return d == nil || (len(d.Xdiff) == 0 && len(d.Ydiff) == 0)
+}
+
+// --------------- BUILD THE TYPE TREE --------------- //
+
+// Targets returns a slice of Targets, one for each child of the Type.
+// In the case of structs, the order is the same as the order of the fields in the struct.
+// In the case of maps, the key Target precedes the value Target.
+func (t *Type) Targets() []Target {
+	targets := make([]Target, 0, len(t.Edges))
+	for _, edge := range t.Edges {
+		targets = append(targets, Target{edge, (*t.children)[edge.String()]})
+	}
+	return targets
+}
+
+// IsLeaf returns true if the Type has no children.
+func (t *Type) IsLeaf() bool {
+	return t.children == nil
```

Comment:
nit: presumably what you have here is fine, but could there be an empty slice?
```suggestion
	return len(t.children) == 0
```


This one might fail on an empty struct...
```suggestion
	return t.Type.Type == 0
```

Reply:
b3280ca354fa5dbda6a7a29b07a4b15eeb440c92